### PR TITLE
feat: Add toast

### LINKED ISSUE #2664: Feedback when copying to clipboard
Would be great to add some kind of toast when we do shift-option-c to copy to clipboard. It's not clear that it's working right now.

### PR DESCRIPTION
**Pull request by h7y**
A simple toast component that can display one toast at a time.

Fixes #2664

### CODE REVIEW

**Review by lipis — CHANGES_REQUESTED** [review]
I think it would be better to show it in the bottom middle of the screen.. and instead of sliding it.. making it appear and disappear (opacity?). It should also stay awake if user has a mouse over it.. once the mouse is out or not there at all.. should go away.

**Comment by h7y** [comment]
I'll do it :) Are the colors and size fine?

**Comment by dwelle** [comment]
I think it should be much smaller. Something like what Google Drive is using?

image

Also, for this particular case of copying PNG to clipboard, make sure the toast isn't displayed on error.

**Comment by lipis** [comment]
> Are the colors and size fine?

we can tweak as we go.

**Comment by h7y** [comment]
I've made the changes 
Demo 2

File: src/locales/en.json [hunk 1]
```diff
@@ -232,5 +232,8 @@
     "title": "Stats for nerds",
     "total": "Total",
     "width": "Width"
+  },
+  "toasts": {
+    "copyToClipboardAsPng": "Copied to clipboard as PNG!"
```
Comment (lipis) [inline src/locales/en.json:237]:
```suggestion
    "copyToClipboardAsPng": "Copied to clipboard as PNG."
```

File: src/components/Toast.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+}) => {
+  const [shouldClear, setShouldClear] = useState(true);
+
+  useEffect(() => {
+    if (message !== null) {
+      const timeout = setTimeout(() => {
+        if (shouldClear) {
+          clearToast();
+        }
+      }, 2000);
```
Comment (lipis) [inline src/components/Toast.tsx:19]:
```suggestion
      }, TOAST_TIMEOUT);
```

and add it in `constatnts.ts`

File: src/components/Toast.scss [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  .toast {
+    box-sizing: border-box;
+    position: fixed;
+    bottom: 10px;
+    left: 50%;
+    margin-left: -150px;
+    z-index: 999999;
+    animation: fade-in 0.5s;
+    background-color: $oc-gray-6;
+    padding: 0.2rem 0.8rem;
```
Comment (lipis) [inline src/components/Toast.scss:13]: I don't think we use rem anywhere else.. can we stick to pixels?

File: src/components/Toast.scss [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    border-radius: 4px;
+    width: 300px;
```
Comment (lipis) [inline src/components/Toast.scss:15]: Maybe?

```suggestion
    width: 300px;
    text-align: center;
```

File: src/components/Toast.scss [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+.excalidraw {
+  .toast {
+    box-sizing: border-box;
+    position: fixed;
+    bottom: 10px;
+    left: 50%;
+    margin-left: -150px;
+    z-index: 999999;
+    animation: fade-in 0.5s;
+    background-color: $oc-gray-6;
```
Comment (lipis) [inline src/components/Toast.scss:12]:
```suggestion
    background-color: $oc-gray-5;
```

File: src/components/Toast.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  const handleMouseLeave = (
+    event: React.MouseEvent<HTMLDivElement, MouseEvent>,
+  ) => {
+    clearToast();
+    setShouldClear(true);
+  };
+
+  return message ? (
+    <div
+      className="toast"
```
Comment (lipis) [inline src/components/Toast.tsx:33]:
```suggestion
      className="Toast"
```

we usually use the component's names

File: src/components/Toast.scss [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    animation: fade-in 0.5s;
+    background-color: $oc-gray-5;
+    padding: 4px 0;
+    border-radius: 4px;
+    width: 300px;
+    text-align: center;
+  }
+
+  .Toast__message {
+    font-size: 1rem;
```
Comment (lipis) [inline src/components/Toast.scss:20]: is that needed? also rem.

File: src/components/Toast.scss [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    bottom: 10px;
+    left: 50%;
+    margin-left: -150px;
+    z-index: 999999;
+    animation: fade-in 0.5s;
+    background-color: $oc-gray-5;
+    padding: 4px 0;
+    border-radius: 4px;
+    width: 300px;
+    text-align: center;
```
Comment (lipis) [inline src/components/Toast.scss:16]:
```suggestion
    text-align: center;
    cursor: pointer;
```

**Comment by lipis** [comment]
While we are on it.. is there anything we could show as a toast?

File: src/components/Toast.scss (hunk 7 above)
Comment (h7y) [reply src/components/Toast.scss]: Just realized that it is redundant. Removed now!

**Comment by lipis** [comment]
Can we also make it to fade out as well on close?

File: src/locales/en.json [hunk 9]
```diff
@@ -232,5 +232,8 @@
     "title": "Stats for nerds",
     "total": "Total",
     "width": "Width"
+  },
+  "toasts": {
```
Comment (lipis) [inline src/locales/en.json:236]:
```suggestion
  "toast": {
```

File: src/components/App.tsx [hunk 10]
```diff
@@ -911,6 +913,7 @@ class App extends React.Component<ExcalidrawProps, AppState> {
         this.canvas!,
         this.state,
       );
+      this.setState({ toastMessage: t("toasts.copyToClipboardAsPng") });
```
Comment (lipis) [inline src/components/App.tsx:916]:
```suggestion
      this.setState({ toastMessage: t("toast.copyToClipboardAsPng") });
```

**Review by lipis — APPROVED** [review]

**Comment by dwelle** [comment]
Btw, we should make the toast use different style in dark-mode.

In fact, I'd align the style to what we use for the "scroll to content" message:

image

image

**Comment by lipis** [comment]
Nice one with the colors!

File: src/components/Toast.scss [hunk 11]
```diff
@@ -0,0 +1,32 @@
+@import "../css/_variables";
+
+.excalidraw {
+  .Toast {
+    animation: fade-in 0.5s;
+    background-color: var(--button-gray-1);
+    border-radius: 4px;
+    bottom: 10px;
+    box-sizing: border-box;
+    cursor: pointer;
```
Comment (lipis) [inline src/components/Toast.scss:10]:
```suggestion
    cursor: default;
```

**Comment by lipis** [comment]
What about adding that when copying styles as well?

**Comment by lipis** [comment]
But in general this is ready for merge.

**Comment by dwelle** [comment]
I wouldn't display it during regular copy. You don't really need a confirmation for such a common action.

**Comment by lipis** [comment]
Not a regular copy of course.. `Cmd+Option+C`

**Comment by dwelle** [comment]
I need to read more careful :smile:

**Comment by h7y** [comment]
I'll also show on pasting styles?

**Comment by dwelle** [comment]
No need on paste IMO — most of the time it should be obvious from the action itself.

File: src/locales/en.json [hunk 12]
```diff
@@ -232,5 +232,9 @@
     "title": "Stats for nerds",
     "total": "Total",
     "width": "Width"
+  },
+  "toast": {
+    "copyToClipboardAsPng": "Copied to clipboard as PNG.",
+    "copyStyles": "Copied styles."
```
Comment (lipis) [inline src/locales/en.json:238]:
```suggestion
    "copyStyles": "Copied styles.",
    "copyToClipboardAsPng": "Copied to clipboard as PNG."
```

**Comment by dwelle** [comment]
I've pushed a change that makes the Toast render conditionally, simplified the timer logic and handlers, and made the Toast reset the timer on `mouseleave` instead of hiding the toast immediately.

**Comment by dwelle** [comment]
And finally, made exporting empty scene throw instead of alerting. This way the Toast isn't shown in this case. Throwing is also more correct.

With this, we're ready to merge.

**Review by dwelle — APPROVED** [review]

**Comment by h7y** [comment]
> I've pushed a change that makes the Toast render conditionally, simplified the timer logic and handlers, and made the Toast reset the timer on `mouseleave` instead of hiding the toast immediately.

Much cleaner!

**Pull request merged**